From 55ce4a9210c055a694af75e142cfeb249d4ed9c4 Mon Sep 17 00:00:00 2001
From: Jiebing Li <jiebing.li@intel.com>
Date: Thu, 11 Dec 2014 15:24:43 +0800
Subject: [PATCH 1/1] usb/gadget: fix kernel oops issue for super speed
 connection

Some of gadget functions do not provide super speed descriptors,
so their ss_descriptor pointer is NULL which will trigger kernel
oops when this pointer is accessed during super speed enumeration.

Change-Id: I8f7c81dbb524f5254f6795d14ddf122e9614f12b
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-9181
Signed-off-by: Jiebing Li <jiebing.li@intel.com>
---
 drivers/usb/gadget/composite.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/drivers/usb/gadget/composite.c b/drivers/usb/gadget/composite.c
index 7ba79e4..8c9bbc8 100644
--- a/drivers/usb/gadget/composite.c
+++ b/drivers/usb/gadget/composite.c
@@ -661,6 +661,17 @@ static int set_config(struct usb_composite_dev *cdev,
 			descriptors = f->fs_descriptors;
 		}
 
+		/*
+		 * some gadget functions don't support super speed mode, so they
+		 * don't provide ss descriptors.
+		 */
+		if (!descriptors) {
+			INFO(cdev, "function %s doesn't support %s\n",
+				f->name, usb_speed_string(gadget->speed));
+			cdev->config = NULL;
+			return -ENODEV;
+		}
+
 		for (; *descriptors; ++descriptors) {
 			struct usb_endpoint_descriptor *ep;
 			int addr;
-- 
1.9.1

